Implemented support for ordering by functions

FlipEverything opened the pull request:
It would be really nice if you could order by a built-in (or user-defined) driver function. Currently ```count()``` and ```group()``` only works with fields specified in the entity's definition because ```fieldAliasMapping()``` appends the tableName before the fieldName. If the fieldName happens to be a function then it does not work. I implemented a simple "function detector" that works with one parameter (mysql) functions. It does not work with complex functions.

This way you can do something like this:
```php
$mapper->all()->order(['WEEK(date)' => 'ASC', 'priority' => 'DESC']);
```
It translates to this:
```mysql
ORDER BY WEEK(table.date) ASC, priority DESC
```
instead of 
```mysql
ORDER BY table.WEEK(date) ASC, priority DESC
```
This is 100% backwards compatible and it does not break the field alias support. What do you guys think?

vlucas commented on tests/FieldAlias.php:
```diff
@@ -71,6 +71,10 @@ public function testLegacyOrderBy()
         $mapper = test_spot_mapper('SpotTest\Entity\Legacy');
+        // Currently it works with mysql only
+        if (!$mapper->connectionIs('mysql')) {
+            $this->markTestSkipped('Not supported in Sqlite nor Postgres.');
```
Why is this MySQL only? Do SQLite and PostgreSQL have issues supporting this?

FlipEverything replied:
They have issues supporting the ```WEEK()``` function. I only tested for mysql but it is indeed bad design.

vlucas commented on tests/FieldAlias.php:
```diff
@@ -87,6 +91,10 @@ public function testLegacyGroupBy()
         $mapper = test_spot_mapper('SpotTest\Entity\Legacy');
+        // Currently it works with mysql only
+        if (!$mapper->connectionIs('mysql')) {
+            $this->markTestSkipped('Not supported in Sqlite nor Postgres.');
```
Spot supports more than just 3 databases. These messages should be specific to MySQL only if the feature is only supported by MySQL. Something like `This feature is only supported in MySQL`.

vlucas commented on tests/FieldAlias.php:
```diff
@@ -66,6 +66,18 @@ public function testLegacyOrderBy()
+    {
+        $mapper = test_spot_mapper('SpotTest\Entity\Legacy');
+        // Currently it works with mysql only
+        if (!$mapper->connectionIs('mysql')) {
```
Is there some generic ANSI-SQL function you can use instead that will work on all supported/tested database types?

FlipEverything commented:
I improved the "function detector" to work with complex queries. I changed the basic tests to be platform independent and added complex function tests for the three tested drivers.

FlipEverything commented:
summon @vlucas

FlipEverything commented:
ping

vlucas commented:
Sorry I was in Montivideo, Uruguay on a work trip for a week. I am back now :)

FlipEverything commented:
Thank you! 
Sorry, I did not mean to bother you.
